tf/utils.go

```diff
@@ -136,6 +138,136 @@ func NewPassReport(plan *tfjson.Plan) types.PassReport {
 	return out
 }
 
+func NewCoverageReportFromState(state *tfjson.State) (types.CoverageReport, error) {
+	out := types.CoverageReport{
+		Coverages: make(map[types.Resource]*coverage.Model, 0),
+	}
+	if state == nil || state.Values == nil || state.Values.RootModule == nil || state.Values.RootModule.Resources == nil {
+		return out, nil
+	}
+	for _, res := range state.Values.RootModule.Resources {
+		if !strings.HasPrefix(res.Address, "azapi_") {
+			continue
+		}
+
+		id := ""
+		if v, ok := res.AttributeValues["id"]; ok {
+			id = v.(string)
+		}
+		resourceType := ""
+		if v, ok := res.AttributeValues["type"]; ok {
+			resourceType = v.(string)
+		}
+
+		body := map[string]interface{}{}
+		if bodyRaw, ok := res.AttributeValues["body"].(string); ok {
+			if value, ok := res.AttributeValues["tags"]; ok && value != nil && len(value.(map[string]interface{})) > 0 {
+				tagsModel := value.(map[string]interface{})
+				if len(tagsModel) != 0 {
+					body["tags"] = tagsModel
+				}
+			}
+			if value, ok := res.AttributeValues["location"]; ok && value != nil && value.(string) != "" {
+				body["location"] = value.(string)
+			}
+
+			if value, ok := res.AttributeValues["identity"]; ok && value != nil && len(value.([]interface{})) > 0 {
+				body["identity"] = expandIdentity(value.([]interface{}))
+			}
+
+			err := json.Unmarshal([]byte(bodyRaw), &body)
+			if err != nil {
+				return out, err
+			}
+		}
+
+		err := out.AddCoverageFromState(id, resourceType, res.Address, body)
+		if err != nil {
+			return out, err
+		}
+	}
+	return out, nil
+}
+
+func NewCoverageReport(plan *tfjson.Plan) (types.CoverageReport, error) {
+	out := types.CoverageReport{
+		Coverages: make(map[types.Resource]*coverage.Model, 0),
+	}
+	if plan != nil {
+		for _, resourceChange := range plan.ResourceChanges {
+			if !strings.HasPrefix(resourceChange.Address, "azapi_") {
+				continue
+			}
+			if resourceChange == nil || resourceChange.Change == nil {
+				continue
+			}
+			if len(resourceChange.Change.Actions) == 1 && resourceChange.Change.Actions[0] == tfjson.ActionNoop {
+				beforeMap, beforeMapOk := resourceChange.Change.Before.(map[string]interface{})
+				if !beforeMapOk {
+					continue
+				}
+				id := ""
+				if v, ok := beforeMap["id"]; ok {
+					id = v.(string)
+				}
+				resourceType := ""
+				if v, ok := beforeMap["type"]; ok {
+					resourceType = v.(string)
+				}
+
+				body := map[string]interface{}{}
+				if bodyRaw, ok := beforeMap["body"].(string); ok {
```

Comment:
>if [](http://example.com/codeflow?start=4&length=2)

use a new function to define this branch ?

Reply:
Thanks, I have changed the code.